File: BSharp/Data/Queries/FilterExpression.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace BSharp.Data.Queries
{
    /// <summary>
    /// Represents a filter argument which is a tree of ANDs (conjunctions), ORs (disjunctions)
    /// and NOTs (negations) with brackets to override the default precedence of these operators.
    /// For example: "(Order/Total gt 1000) and (Customer/Gender eq 'M')"
    /// This class also contains utilities for parsing filter arguments into <see cref="FilterExpression"/>s.
    /// IMPORTANT: there is a high fidality replica of this in TypeScript for the ClientApp , the two must be kept in sync
    /// </summary>
    public abstract class FilterExpression : IEnumerable<FilterAtom>
    {
        /// <summary>
        /// Parses a string representing a filter argument into a <see cref="FilterExpression"/>. 
        /// The filter argument is a tree of ANDs (conjunctions), ORs (disjunctions)
        /// and NOTs (negations) with brackets to override the default precedence of these operators.
        /// For example "(Order/Total gt 1000) and (Customer/Gender eq 'M')"
        /// </summary>
        public static FilterExpression Parse(string filter)
        {
            if (string.IsNullOrWhiteSpace(filter))
            {
                return null;
            }

            string preprocessedFilter = Preprocess(filter);
            IEnumerable<string> tokenStream = Tokenize(preprocessedFilter);
            FilterExpression filterExpression = ParseTokenStream(tokenStream);

            return filterExpression;
        }

        /// <summary>
        /// Preprocesses the filer expression: trimming it and removing any repeated spaces
        /// </summary>
        private static string Preprocess(string filter)
        {
            // Ensure no spaces are repeated
            Regex regex = new Regex("[ ]{2,}", RegexOptions.None);
            filter = regex.Replace(filter, " ");

            // Trim
            filter = filter.Trim();

            // return preprocessed filter argument
            return filter;
        }

        /// <summary>
        /// Lexical Analysis: Turns the filter expression into a stream of recognized tokens
        /// </summary>
        private static IEnumerable<string> Tokenize(string preprocessedFilter)
        {
            // All the symbols, we use spaces before and after the operators to tell them apart
            // Special handling for "not(E)" we need to allow for no space before or after it, and at
            // the same time cannot get confused with property names like "Notes"
            List<string> symbols = new List<string>(new string[] { // the order matters

                    // Logical Operators
                    " and ", " or ", "not",
                    
                    // Parentheses
                    "(", ")",
                });

            // For performance: decompose the filter into a char array and use a string builder to accumulate the characters examined so far
            char[] filterArray = preprocessedFilter.ToCharArray();
            bool insideQuotes = false;
            StringBuilder acc = new StringBuilder();            
            int index = 0;

            string MatchingOperator(int i)
            {
                // This basically finds the first symbol that matches the beginning of the current index at filterArray
                var matchingSymbol = symbols.FirstOrDefault(symbol => (filterArray.Length - i) >= symbol.Length && 
                    Enumerable.Range(0, symbol.Length).All(j => char.ToLower(symbol[j]) == char.ToLower(filterArray[i + j])));

                if(matchingSymbol == "not")
                {
                    // The operator "not" requires more elaborate handling, since it may not necessarily be preceded or superseded by a space
                    // but we don't want to confuse it with properties that contain "not" in their name like "Notes"
                    int prevIndex = i - 1;
                    bool precededProperly = prevIndex < 0 || filterArray[prevIndex] == ' ' || filterArray[prevIndex] == '(';
                    int nextIndex = i + matchingSymbol.Length;
                    bool followedProperly = nextIndex >= filterArray.Length || filterArray[nextIndex] == ' ' || filterArray[nextIndex] == '(';

                    return precededProperly && followedProperly ? matchingSymbol : null;
                }
                else
                {
                    // All the other symbols can be precisely matched
                    return matchingSymbol;
                }
            }

            while (index < filterArray.Length)
            {
                bool isSingleQuote = filterArray[index] == '\'';

                if (isSingleQuote)
                {
                    bool followedBySingleQuote = (index + 1) < filterArray.Length && filterArray[index + 1] == '\'';

                    acc.Append(filterArray[index]);
                    index++;

                    if (!insideQuotes)
                    {
                        insideQuotes = true;
                    }
                    else if (!followedBySingleQuote)
                    {
                        insideQuotes = false;
                    }
                    else // inside quotes and followed by a single quote
                    {
                        index++; // skip the other single quote, it's just there for escaping the first one
                    }
                }
                else
                {
                    // Everything that is not inside single quotes is ripe for lexical analysis   
                    string matchingSymbol;
                    if (!insideQuotes && (matchingSymbol = MatchingOperator(index)) != null)
                    {
                        // Add all that has been accumulating before the symbol
                        if (acc.Length > 0)
                        {
                            var token = acc.ToString().Trim();
                            if (!string.IsNullOrWhiteSpace(token))
                            {
                                yield return token;
                            }

                            acc = new StringBuilder();
                        }

                        // And add the symbol  
                        yield return matchingSymbol.ToLower().Trim();
                        index += matchingSymbol.Length;
                    }
                    else
                    {
                        acc.Append(filterArray[index]);
                        index++;
                    }
                }
            }

            if (insideQuotes)
            {
                // Programmer mistake
                throw new InvalidOperationException("Uneven number of single quotation marks in filter query parameter, quotation marks in literals should be escaped by specifying them twice");
            }

            if (acc.Length > 0)
            {
                var token = acc.ToString().Trim();
                if (!string.IsNullOrWhiteSpace(token))
                {
                    yield return token;
                }
            }
        }

        /// <summary>
        /// Constructs and returns an abstract expression tree (<see cref="FilterExpression"/>) based on a token stream
        /// </summary>
        public static FilterExpression ParseTokenStream(IEnumerable<string> tokens)
        {
            // This is an implementation of the shunting-yard algorithm from Edsger Dijkstra https://bit.ly/1fEvvLI

            var ops = new Stack<string>();
            var output = new Stack<FilterExpression>();

            // Inline function to make it easy to add tokens to the output stack
            void AddToOutput(string token)
            {
                switch (token)
                {
                    case "and":
                        if (output.Count < 2)
                        {
                            throw new InvalidOperationException("Badly formatted filter parameter, a conjunction 'and' was missing one or both of its 2 operands");
                        }

                        output.Push(FilterConjunction.Make(left: output.Pop(), right: output.Pop()));
                        break;
                    case "or":
                        if (output.Count < 2)
                        {
                            throw new InvalidOperationException("Badly formatted filter parameter, a disjunction 'or' was missing one or both of its 2 operands");
                        }

                        output.Push(FilterDisjunction.Make(left: output.Pop(), right: output.Pop()));
                        break;
                    case "not":
                        if (output.Count < 1)
                        {
                            throw new InvalidOperationException("Badly formatted filter parameter, a negation 'not' was missing its operand");
                        }

                        output.Push(FilterNegation.Make(inner: output.Pop()));
                        break;
                    default:
                        output.Push(FilterAtom.Parse(token));
                        break;
                }
            }

            // The shunting-yard implementation
            foreach (var token in tokens)
            {
                if (token == "not")
                {
                    // if it is a logical negation push it on the operators stack
                    ops.Push(token);
                }
                else if (_operators.TryGetValue(token, out OperatorInfo opInfo)) // if it is an operator
                {
                    // inline predicate determines how many items do we pop from the operator stack
                    bool KeepPopping()
                    {
                        /* Modified from Wikipedia: Keep popping while...
                          
                            (the operator at the top of the operator stack is not a left parenthesis) AND 
                            (   
                                (there is a 'not' at the top of the operator stack) OR
                                (there is an operator at the top of the operator stack with greater precedence) OR
                                (the operator at the top of the operator stack has equal precedence and is left associative)
                            )
                         */

                        if (ops.Count == 0)
                        {
                            // There is nothing left to pop
                            return false;
                        }

                        string peek = ops.Peek();
                        _operators.TryGetValue(peek, out OperatorInfo peekInfo);

                        return peek != "(" &&
                            (
                                peek == "not" || peekInfo.Precedence > opInfo.Precedence ||
                                (peekInfo.Precedence == opInfo.Precedence && peekInfo.IsLeftAssociative)
                            );
                    }

                    while (KeepPopping())
                    {
                        AddToOutput(ops.Pop());
                    }

                    ops.Push(token);
                }
                else if (token == "(")
                {
                    ops.Push(token);
                }
                else if (token == ")")
                {
                    if (ops.Count == 0)
                    {
                        // There should have been a left paren in the stack
                        throw new InvalidOperationException("Filter expression contains mismatched parentheses");
                    }

                    // Keep popping from the operator queue until you hit a left paren
                    while (ops.Count > 0 && ops.Peek() != "(")
                    {
                        // Add to output
                        AddToOutput(ops.Pop());
                    }

                    if (ops.Count > 0 && ops.Peek() == "(")
                    {
                        // Pop the left paren
                        ops.Pop();
                    }
                    else
                    {
                        // There should have been a left paren in the stack
                        throw new InvalidOperationException("Filter expression contains mismatched parentheses");
                    }
                }
                else
                {
                    // It's a simple atom, add it the output
                    AddToOutput(token);
                }
            }

            // Anything left in the operators queue, add it to the output
            while (ops.Count > 0)
            {
                if (ops.Peek() != "(")
                {
                    // Add to output
                    AddToOutput(ops.Pop());
                }
                else
                {
                    // Depends whether you want to be forgiving of left parentheses that weren't closed
                    // ops.Pop();

                    // There should not be a left paren in the stack
                    throw new InvalidOperationException("Filter expression contains mismatched parentheses");
                }
            }

            // If the filter expression is valid, there should be exactly one item in the output stack at this stage
            if (output.Count != 1)
            {
                throw new InvalidOperationException("Badly formatted filter parameter");
            }

            return output.Pop();
        }

        /// <summary>
        /// Returns all the <see cref="FilterAtom"/> in this current expression tree
        /// </summary>
        /// <returns></returns>
        public abstract IEnumerable<FilterAtom> Atoms();

        /// <summary>
        /// Implementation of <see cref="IEnumerable{T}"/>
        /// </summary>
        public IEnumerator<FilterAtom> GetEnumerator()
        {
            return Atoms().GetEnumerator();
        }

        /// <summary>
        /// Implementation of <see cref="IEnumerable{T}"/>
        /// </summary>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return Atoms().GetEnumerator();
        }

        /// <summary>
        /// This list contains the precedence and associativity of supported binary operators, the precedences used are the same as T-SQL (https://bit.ly/2YnyfbV)
        /// </summary>
        private static readonly Dictionary<string, OperatorInfo> _operators = new Dictionary<string, OperatorInfo>
        {
            ["and"] = new OperatorInfo { Precedence = 6, Associativity = Associativity.Left },
            ["or"] = new OperatorInfo { Precedence = 7, Associativity = Associativity.Left },
        };

        /// <summary>
        /// Used internally to store the precedence and associativity of a binary operator
        /// </summary>
        private struct OperatorInfo
        {
            public int Precedence { get; set; }

            public Associativity Associativity { get; set; }

            public bool IsLeftAssociative
            {
                get
                {
                    return Associativity == Associativity.Left;
                }
            }
        }

        /// <summary>
        /// https://bit.ly/2Kp2Yvl
        /// </summary>
        private enum Associativity { Left, Right }
    }
}

File: BSharp/ClientApp/src/app/data/filter-expression.ts
/**
 * The base class for all filter expressions
 */
export abstract class FilterExpressionBase {
    public abstract toString(): string;
}

/**
 * Represents a logical disjunction (and)
 */
export class FilterDisjunction extends FilterExpressionBase {
    type: 'disjunction';
    left: FilterExpression;
    right: FilterExpression;

    public toString(): string {
        return `(${this.left.toString()}) or (${this.right.toString()})`;
    }
}

/**
 * Represents a logical disjunction (or)
 */
export class FilterConjunction extends FilterExpressionBase {
    type: 'conjunction';
    left: FilterExpression;
    right: FilterExpression;

    public toString(): string {
        return `(${this.left.toString()}) and (${this.right.toString()})`;
    }
}

/**
 * Represents a logical negation (not)
 */
export class FilterNegation extends FilterExpressionBase {
    type: 'negation';
    inner: FilterExpression;

    public toString(): string {
        return `not(${this.inner.toString()})`;
    }
}

/**
 * Represents a filter atom (e.g. Code eq '11')
 */
export class FilterAtom extends FilterExpressionBase {
    type: 'atom';
    path: string[];
    property: string;
    op: string;
    value: string;

    public static parse(atom: string): FilterAtom {
        const pieces = atom.split(' ').filter(e => !!e);
        if (pieces.length < 3) {
            throw new Error(`One of the atomic expressions (${atom}) does not have the valid form: 'Path op Value'`);
        } else {
            // get the path and the prop
            const steps = pieces.shift().split('/').map(e => !e ? '' : e.trim());
            const property = steps.pop();
            const path = steps;

            // get the op
            const op = pieces.shift();

            // get the value
            const value = pieces.join(' ').trim();

            return {
                type: 'atom',
                path,
                property,
                op,
                value
            };
        }
    }

    public toString(): string {
        const stringPath = this.path.concat([this.property]).join('/');
        return `${stringPath} ${this.op} ${this.value}`;
    }
}

/**
 * The union of all Filter Expressions
 */
export declare type FilterExpression = FilterDisjunction | FilterConjunction | FilterNegation | FilterAtom;

export class OperatorInfo {
    constructor(public precedence: number, public associativity: 'left' | 'right') { }

    public get isLeftAssociative(): boolean {
        return this.associativity === 'left';
    }
}

export class FilterTools {
    private static _operators: { [op: string]: OperatorInfo } = {
        and: new OperatorInfo(6, 'left'),
        or: new OperatorInfo(7, 'left'),
    };

    public static areEquivalent(exp1: FilterExpression, exp2: FilterExpression) {
        if (exp1 === exp2) {
            return true;
        } else if (!exp1) {
            return !exp2;
        } else if (!exp2) {
            return !exp1;
        } else {
            switch (exp1.type) {
                case 'conjunction':
                    return exp2.type === 'conjunction' && FilterTools.areEquivalent(exp1.left, exp2.left)
                        && FilterTools.areEquivalent(exp1.right, exp2.right);

                case 'disjunction':
                    return exp2.type === 'disjunction' && FilterTools.areEquivalent(exp1.left, exp2.left)
                        && FilterTools.areEquivalent(exp1.right, exp2.right);

                case 'negation':
                    return exp2.type === 'negation' && FilterTools.areEquivalent(exp1.inner, exp2.inner);

                case 'atom':
                    return exp2.type === 'atom' && exp1.value === exp2.value;
            }
        }
    }

    public static placeholderAtoms(exp: FilterExpression): FilterAtom[] {
        if (!exp) {
            return [];
        }

        switch (exp.type) {
            case 'conjunction':
            case 'disjunction':
                return FilterTools.placeholderAtoms(exp.left).concat(FilterTools.placeholderAtoms(exp.right));
            case 'negation':
                return FilterTools.placeholderAtoms(exp.inner);
            case 'atom':
                return !!exp.value && exp.value.startsWith('@') ? [exp] : [];
        }

        return [];
    }

    /**
     * Stringifies a FilterExpression into an odata like query string
     * @param exp the FilterExpression to stringify
     */
    public static stringify(exp: FilterExpression): string {
        if (!exp) {
            return null;
        }

        switch (exp.type) {
            case 'conjunction': {
                // only use parentheses if the child is a disjunction
                const leftRaw = FilterTools.stringify(exp.left);
                const left = exp.left.type === 'disjunction' ? `(${leftRaw})` : leftRaw;
                const rightRaw = FilterTools.stringify(exp.right);
                const right = exp.right.type === 'disjunction' ? `(${rightRaw})` : rightRaw;

                return `${left} and ${right}`;
            }
            case 'disjunction': {
                // only use parentheses if the child is a conjunction
                const leftRaw = FilterTools.stringify(exp.left);
                const left = exp.left.type === 'conjunction' ? `(${leftRaw})` : leftRaw;
                const rightRaw = FilterTools.stringify(exp.right);
                const right = exp.right.type === 'conjunction' ? `(${rightRaw})` : rightRaw;

                return `${left} or ${right}`;
            }
            case 'negation': {
                const inner = FilterTools.stringify(exp.inner);
                return `not(${inner})`;
            }
            case 'atom': {
                const stringPath = exp.path.concat([exp.property]).join('/');
                return `${stringPath} ${exp.op} ${exp.value}`;
            }
        }
    }

    /**
     * Parses an Odata-like filter string into a FilterExpression tree
     * @param filter the filter string to parse
     */
    public static parse(filter: string): FilterExpression {

        const preprocessedFilter = this.preprocess(filter);
        if (!preprocessedFilter) {
            return null;
        }

        const tokenStream = this.tokenize(preprocessedFilter);
        const filterExpression = this.parseTokenStream(tokenStream);

        return filterExpression;
    }

    private static preprocess(filter: string): string {
        filter = filter || '';

        // Reduce all enters to a single space
        let oldFilter: string = null;
        while (oldFilter !== filter) {
            oldFilter = filter;
            filter = filter.replace(/(?:\r\n|\r|\n)/g, ' ');
        }

        // Reduce all double spaces to a single space
        oldFilter = null;
        while (oldFilter !== filter) {
            oldFilter = filter;
            filter = filter.replace('  ', ' ');
        }

        filter = filter.trim();
        return filter;
    }

    private static tokenize(prerocessedFilter: string): string[] {
        const symbols = [' and ', ' or ', 'not', '(', ')'];
        const filterArray = prerocessedFilter.split('');
        let insideQuotes = false;
        let acc: string[] = [];
        let index = 0;
        const result: string[] = [];

        function matchingOperator(i: number): string {
            const matchingSymbol = symbols.find(symbol => (filterArray.length - i) >= symbol.length &&
                Array.from(Array(symbol.length).keys()).every(j => symbol[j].toLowerCase() === filterArray[i + j].toLowerCase()));

            if (matchingSymbol === 'not') {
                const prevIndex = i - 1;
                const precededProperly = prevIndex < 0 || filterArray[prevIndex] === ' ' || filterArray[prevIndex] === '(';
                const nextIndex = i + matchingSymbol.length;
                const followedProperly = nextIndex >= filterArray.length || filterArray[nextIndex] === ' '
                    || filterArray[nextIndex] === '(';

                return precededProperly && followedProperly ? matchingSymbol : null;
            } else {
                return matchingSymbol;
            }
        }

        while (index < filterArray.length) {
            const isSingleQuote = filterArray[index] === '\'';

            if (isSingleQuote) {
                const followedBySingleQuote = (index + 1) < filterArray.length && filterArray[index + 1] === '\'';

                acc.push(filterArray[index]);
                index++;

                if (!insideQuotes) {
                    insideQuotes = true;
                } else if (!followedBySingleQuote) {
                    insideQuotes = false;
                } else {
                    index++;
                }
            } else {
                let matchingSymbol: string;
                // tslint:disable:no-conditional-assignment
                if (!insideQuotes && !!(matchingSymbol = matchingOperator(index))) {
                    if (acc.length > 0) {
                        const token = acc.join('').trim();
                        if (!!token) {
                            result.push(token);
                        }
                        acc = [];
                    }

                    result.push(matchingSymbol.toLowerCase().trim());
                    index = index + matchingSymbol.length;
                } else {
                    acc.push(filterArray[index]);
                    index++;
                }
            }
        }

        if (insideQuotes) {
            throw new Error(`Uneven number of single quotation marks in filter query parameter,
            quotation marks in literals should be escaped by specifying them twice`);
        }

        if (acc.length > 0) {
            const token = acc.join('').trim();
            if (!!token) {
                result.push(token);
            }
        }

        return result;
    }

    private static parseTokenStream(tokens: string[]): FilterExpression {

        const ops: string[] = [];
        const output: FilterExpression[] = [];

        function addToOutput(token: string) {
            switch (token) {
                case 'and':
                    if (output.length < 2) {
                        throw new Error(`Badly formatted filter parameter, a conjunction 'and' was missing one or both of its 2 operands`);
                    }

                    output.push({ type: 'conjunction', left: output.pop(), right: output.pop() });
                    break;
                case 'or':
                    if (output.length < 2) {
                        throw new Error(`Badly formatted filter parameter, a disjunction 'or' was missing one or both of its 2 operands`);
                    }

                    output.push({ type: 'disjunction', left: output.pop(), right: output.pop() });
                    break;
                case 'not':
                    if (output.length < 1) {
                        throw new Error(`Badly formatted filter parameter, a negation 'not' was missing its operand`);
                    }

                    output.push({ type: 'negation', inner: output.pop() });
                    break;
                default:
                    output.push(FilterAtom.parse(token));
                    break;
            }
        }

        for (const token of tokens) {

            if (token === 'not') {
                ops.push(token);
            } else if (this._operators[token]) {
                const opInfo = this._operators[token];

                const keepPopping: () => boolean = () => {
                    if (ops.length === 0) {
                        return false;
                    }

                    const peek = ops[ops.length - 1];
                    const peekInfo = this._operators[peek];

                    return peek !== '(' &&
                        (
                            peek === 'not' || peekInfo.precedence > opInfo.precedence ||
                            (peekInfo.precedence === opInfo.precedence && peekInfo.isLeftAssociative)
                        );
                };

                while (keepPopping()) {
                    addToOutput(ops.pop());
                }

                ops.push(token);
            } else if (token === '(') {
                ops.push(token);
            } else if (token === ')') {
                if (ops.length === 0) {
                    throw new Error(`Filter expression contains mismatched parentheses`);
                }

                while (ops.length > 0 && ops[ops.length - 1] !== '(') {
                    addToOutput(ops.pop());
                }

                if (ops.length > 0 && ops[ops.length - 1] === '(') {
                    ops.pop();
                } else {
                    throw new Error(`Filter expression contains mismatched parentheses`);
                }
            } else {
                addToOutput(token);
            }
        }

        while (ops.length > 0) {
            if (ops[ops.length - 1] !== '(') {
                // Add to output
                addToOutput(ops.pop());
            } else {
                throw new Error(`Filter expression contains mismatched parentheses`);
            }
        }

        if (output.length !== 1) {
            throw new Error(`Badly formatted filter parameter`);
        }

        return output.pop();
    }
}
